src/testRunner/unittests/tsserver/typingsInstaller.ts

```diff
@@ -964,13 +969,49 @@ namespace ts.projectSystem {
 
             checkNumberOfProjects(projectService, { configuredProjects: 1 });
             const p = configuredProjectAt(projectService, 0);
-            checkProjectActualFiles(p, [app.path, jsconfig.path]);
+            const filesBeforeInstall = jQueryJsInProjectBeforeInstall ? [app.path, jquery.path, jsconfig.path] : [app.path, jsconfig.path];
+            checkProjectActualFiles(p, filesBeforeInstall);
 
-            installer.installAll(/*expectedCount*/ 1);
+            installer.installAll(jQueryDtsInProjectAfterInstall ? 1 : 0);
 
             checkNumberOfProjects(projectService, { configuredProjects: 1 });
-            host.checkTimeoutQueueLengthAndRun(2);
-            checkProjectActualFiles(p, [app.path, jqueryDTS.path, jsconfig.path]);
+            host.checkTimeoutQueueLengthAndRun(jQueryDtsInProjectAfterInstall ? 2 : 0);
+            checkProjectActualFiles(p, jQueryDtsInProjectAfterInstall ? [app.path, jqueryDTS.path, jsconfig.path] : filesBeforeInstall);
+        }
+
+        it("configured projects discover from node_modules", () => {
+            testConfiguredProjectNodeModules({
+                jQueryJsInProjectBeforeInstall: false,
+                jQueryDtsInProjectAfterInstall: true,
+            });
+        });
+
+        it("configured projects discover from node_modules - empty types", () => {
+            // Explicit types prevent automatic inclusion from package.json listing
+            testConfiguredProjectNodeModules({
+                jsconfigContent: { compilerOptions: { types: [] } },
+                jQueryJsInProjectBeforeInstall: false,
+                jQueryDtsInProjectAfterInstall: false,
+            });
+        });
+
+        it("configured projects discover from node_modules - explicit types", () => {
+            // A type reference directive will not resolve to the global typings cache

```

Comment:
Should it though especially if `typeRoots` is not specified?

Reply:
Actually if program has "types" and that type is not present it should be part of unresolved import list right ?. Can you add real tsserver end to end test for this so we can verify.

Reply:
I convinced myself it actually should not [here](https://github.com/microsoft/TypeScript/pull/49275#discussion_r889152599):

> In other words, unresolved imports can trigger special resolution behavior because imports first and foremost are written for use at runtime; bringing in the typings on top of them is an added bonus to help your experience at design time. Type reference directives (including `"types"`) are different because they have no runtime purpose; if you wrote them, you were intentionally using a TypeScript-only feature and you should do so completely by installing the types/dependencies you say you want to reference.